### DISCUSSION TITLE
partition.dat

### DISCUSSION THREAD

**Discussion by adelie93** [rgerganov/gen5fw#15] [Q&A]
Can you explain how `partition.dat` is created?

**Comment by rgerganov** [comment]
I don't know how it is created, I found it in on the internet and it worked for me. Any information about its structure is welcome.

**Reply by adelie93** [reply]
This is what I've been able to puzzle out so far.
Using `fdisk` to see the partition table on my device:
```
root@android:/ # busybox fdisk -l /dev/block/mmcblk0

Disk /dev/block/mmcblk0: 7818 MB, 7818182656 bytes
1 heads, 16 sectors/track, 954368 cylinders
Units = cylinders of 16 * 512 = 8192 bytes

              Device Boot      Start         End      Blocks  Id System
/dev/block/mmcblk0p1               3        2050       16384   0 Empty
/dev/block/mmcblk0p2            2051       57346      442368  83 Linux
/dev/block/mmcblk0p3           57347      114690      458752  83 Linux
/dev/block/mmcblk0p4          114691      468992     2834416   f Win95 Ext'd (LBA)
/dev/block/mmcblk0p5          114693      129028      114688  83 Linux
/dev/block/mmcblk0p6          129031      131078       16384   0 Empty
/dev/block/mmcblk0p7          131081      132104        8192   0 Empty
/dev/block/mmcblk0p8          132107      132618        4096   0 Empty
/dev/block/mmcblk0p9          132621      133132        4096   0 Empty
/dev/block/mmcblk0p10         133135      157710      196608   0 Empty
/dev/block/mmcblk0p11         157713      165904       65536  83 Linux
/dev/block/mmcblk0p12         165907      231442      524288  83 Linux
/dev/block/mmcblk0p13         231445      376852     1163264  83 Linux
/dev/block/mmcblk0p14         376855      442390      524288  83 Linux
/dev/block/mmcblk0p15         442393      468992      212800  83 Linux
```
And trying to correlate it to what's in `partition.dat`.
Looking at the first 4 partitions, as this seems similar to a classic MBR partition where to get more than 4 you need an "extended" partition.

start/end are in sectors, sector = 8 blocks
block size = 1024 bytes
fdisk shows size in blocks, the raw partition table is in sectors

looking at the numbers from `fdisk` and converting them to hex, I get:
```
(starting sector)				(size in sectors)
3-1 = 2 = 0x2					16384/8 = 2048 = 0x800
2051-1 = 2050 = 0x802			442368/8 = 55296 = 0xd800
57347-1 = 57346 = 0xe002		458752/8 = 57334 = 0xe000
114691-1 = 114690 = 0x1c002		2834416/8 = 354302 = 0x567fe
```

Now looking at the `partition.dat` file:
```
$ xxd partition.dat | head   
00000000: 0c00 0000 0000 0000 0000 0000 0000 0000  ................
00000010: 0000 0000 2000 0000 0080 0000 0000 0000  .... ...........
00000020: 8300 0000 2080 0000 0080 0d00 0000 0000  .... ...........
00000030: 8300 0000 2000 0e00 0000 0e00 0000 0000  .... ...........
00000040: 0f00 0000 2000 1c00 e07f 5600 0040 0038  .... .....V..@.8
```

Same data but with some annotations I added:
```
           0 1  2 3  4 5  6 7  8 9  a b  c d  e f
00000000: 0c00 0000 0000 0000 0000 0000 0000 0000  ................

(blk0p1)  TYPE UNK_ START_SEC SIZE_SEC_ UNKNOWN__
00000010: 0000 0000 2000 0000 0080 0000 0000 0000  .... ...........

(blk0p2)  TYPE UNK_ START_SEC SIZE_SEC_ UNKNOWN__
00000020: 8300 0000 2080 0000 0080 0d00 0000 0000  .... ...........

(blk0p3)  TYPE UNK_ START_SEC SIZE_SEC_ UNKNOWN__
00000030: 8300 0000 2000 0e00 0000 0e00 0000 0000  .... ...........

(blk0p4)  TYPE UNK_ START_SEC SIZE_SEC_ UNKNOWN__
00000040: 0f00 0000 2000 1c00 e07f 5600 0040 0038  .... .....V..@.8
```

Still lots to understand but that's at least a start...

**Reply by adelie93** [reply]
Looked at an implementation of `fdisk` ([here](https://github.com/brgl/busybox/blob/abbf17abccbf832365d9acf1c280369ba7d5f8b2/util-linux/fdisk.c)) for more information on the meaning of these bytes:
```
struct partition {
	unsigned char boot_ind;         /* 0x80 - active */
	unsigned char head;             /* starting head */
	unsigned char sector;           /* starting sector */
	unsigned char cyl;              /* starting cylinder */
	unsigned char sys_ind;          /* what partition type */
	unsigned char end_head;         /* end head */
	unsigned char end_sector;       /* end sector */
	unsigned char end_cyl;          /* end cylinder */
	unsigned char start4[4];        /* starting sector counting from 0 */
	unsigned char size4[4];         /* nr of sectors in partition */
} PACKED;
```

Based on that, my new understanding of the `partition.dat` file is:
```
          BI HD SC CY SI EH ES EC START_SEC SIZE_SEC_
0000000c: 00 00 00 00 00 00 00 00 2000 0000 0080 0000  ........ .......	p1	mmcblk0.img1            32   32799   32768    16M  0 Empty
0000001c: 00 00 00 00 83 00 00 00 2080 0000 0080 0d00  ........ .......	p2	mmcblk0.img2         32800  917535  884736   432M 83 Linux
0000002c: 00 00 00 00 83 00 00 00 2000 0e00 0000 0e00  ........ .......	p3	mmcblk0.img3        917536 1835039  917504   448M 83 Linux
0000003c: 00 00 00 00 0f 00 00 00 2000 1c00 e07f 5600  ........ .....V.	p4	mmcblk0.img4       1835040 7503871 5668832   2.7G  f W95 Ext'd (LBA)
```

That same data appears in the real filesystem at 00001be.

Understanding that, it's now plain what the structure of `partition.dat` is:
1. First 4 bytes (0x0000000c, decimal 12) are the number of blocks of data that need to be written to disk  
2. Next 8 bytes are the offset on disk to where the next block of data should go (0x00 for this first set of data)
3. Next 64 bytes are the partition data
4. loop back to step 2 until all 12 blocks are written.

So here are the first few blocks of data from `partition.dat` broken apart into the parts and compared between the real filesystem and the partition file.  
address: represents the location of the data in the `partition.dat` file  
(address) represents the location of the data in the final, real filesystem  
```
                      0 1  2 3  4 5  6 7  8 9  a b  c d  e f
           00000000: 0c00 0000                                         ....
---------
           00000004: 0000 0000 0000 0000                               ........
---------
(000001be) 0000000c: 0000 0000 0000 0000 2000 0000 0080 0000  ........ .......	p1	mmcblk0.img1            32   32799   32768    16M  0 Empty
(000001ce) 0000001c: 0000 0000 8300 0000 2080 0000 0080 0d00  ........ .......	p2	mmcblk0.img2         32800  917535  884736   432M 83 Linux
(000001de) 0000002c: 0000 0000 8300 0000 2000 0e00 0000 0e00  ........ .......	p3	mmcblk0.img3        917536 1835039  917504   448M 83 Linux
(000001ee) 0000003c: 0000 0000 0f00 0000 2000 1c00 e07f 5600  ........ .....V.	p4	mmcblk0.img4       1835040 7503871 5668832   2.7G  f W95 Ext'd (LBA)
(000001fe): 55aa
---------
           0000004c: 0040 0038 0000 0000                      .@.8....
---------
                              // start sec is an offset from extended partition start, not disk start
(380041be) 00000054: 0000 0000 8300 0000 2000 0000 0080 0300  ........ .......	p5	mmcblk0.img5       1835072 2064447  229376   112M 83 Linux
(380041ce) 00000064: 0000 0000 0f00 0000 2080 0300 2080 0000  ........ ... ...
(380041de) 00000074: 0000 0000 0000 0000 0000 0000 0000 0000  ................
(380041ee) 00000084: 0000 0000 0000 0000 0000 0000 0000 0000  ................
(380041fe): 55aa
---------
           00000094: 0080 003f 0000 0000                      ...?....
---------
(3f0081be) 0000009c: 0000 0000 0000 0000 2000 0000 0080 0000  ........ .......	p6	mmcblk0.img6       2064480 2097247   32768    16M  0 Empty
(3f0081ce) 000000ac: 0000 0000 0f00 0000 4000 0400 2040 0000  ........@... @..
(3f0081de) 000000bc: 0000 0000 0000 0000 0000 0000 0000 0000  ................
(3f0081ee) 000000cc: 0000 0000 0000 0000 0000 0000 0000 0000  ................
(3f0081fe): 55aa
---------
           000000dc: 00c0 0040 0000 0000                      ...@....
---------
(4000c1be) 000000e4: 0000 0000 0000 0000 2000 0000 0040 0000  ........ ....@..	p7	mmcblk0.img7       2097280 2113663   16384     8M  0 Empty
(4000c1ce) 000000f4: 0000 0000 0f00 0000 6040 0400 2020 0000  ........`@..  ..
(4000c1de) 00000104: 0000 0000 0000 0000 0000 0000 0000 0000  ................
(4000c1ee) 00000114: 0000 0000 0000 0000 0000 0000 0000 0000  ................
(4000c1fe): 55aa
```

I think this fully explains the structure of `partition.dat` now.

**Reply by adelie93** [reply]
Using the information above, I created an edited `partition.dat` file and tried the flashing procedure with it.

Attempt 1: I changed the size of many partitions, attempting to increase the size of `/system` and `/data`, and to use the huge amount of "wasted" space, as the current partitions only use about 3GB of the 8GB eMMC.  
This failed.  
After my edits, the `recovery` partition should have been at block `3412000`. But observing the update process, I saw:
```
Erase eMMC finished!
eMMC encrypt update start !!!
[security_force/encrypt_partition.dat] decrypt&write start[4096 bytes]
[security_force/encrypt_partition.dat] decrypt&write success
[security_force/encrypt_boot.img] decrypt&write start[10579968 bytes] sector[32]
[security_force/encrypt_boot.img] decrypt&write success
[security_force/encrypt_system.img] decrypt&write start[397752112 bytes] sector[32800]
[security_force/encrypt_system.img] decrypt&write success
[security_force/encrypt_recovery.img] decrypt&write start[11554816 bytes] sector[2064480]
[security_force/encrypt_recovery.img] decrypt&write success
[security_force/encrypt_splash.img] decrypt&write start[5745080 bytes] sector[2097280]
[security_force/encrypt_splash.img] decrypt&write success
[security_force/encrypt_lk.rom] decrypt&write start[444332 bytes] sector[0]
[security_force/encrypt_lk.rom] decrypt&write success

Check update option file valid
[security_force/snapshot.sparse.img] file not find

VR image option write check

USB only Encrypt Update SUCCESS

Please disconnect USB connection.
```

Here the `recovery` is still being written to its original location of block `2064480`.  
Therefore I conclude that the values from the upgrade process must be hard-coded, they do not change based on the partitions as defined in `partition.dat`.

So, changing `partition.dat` is unlikely to be very useful, it seems. The locations of `boot`, `system`, `recovery`, `splash`, and `lk` are fixed on the disk and so changing the partitions make them end up in the wrong place. Re-numbering the partitions is likewise untenable as the init scripts all refer to partitions by their partition numbers.

On attempt 2, I went for a much more modest change. I changed only a few partitions that appear after block `2097280`, the last hard-coded reference for the upgrade process.  
This DID work. I successfully shrunk partition 13 and used the extra space to create a new partition that didn't exist previously, partition 16:
```
root@android:/ # busybox fdisk -l /dev/block/mmcblk0

Disk /dev/block/mmcblk0: 7818 MB, 7818182656 bytes
1 heads, 16 sectors/track, 954368 cylinders
Units = cylinders of 16 * 512 = 8192 bytes

              Device Boot      Start         End      Blocks  Id System
/dev/block/mmcblk0p1               3        2050       16384   0 Empty
/dev/block/mmcblk0p2            2051       57346      442368  83 Linux
/dev/block/mmcblk0p3           57347      114690      458752  83 Linux
/dev/block/mmcblk0p4          114691      468992     2834416   f Win95 Ext'd (LBA)
/dev/block/mmcblk0p5          114693      129028      114688  83 Linux
/dev/block/mmcblk0p6          129031      131078       16384   0 Empty
/dev/block/mmcblk0p7          131081      132104        8192   0 Empty
/dev/block/mmcblk0p8          132107      132618        4096   0 Empty
/dev/block/mmcblk0p9          132621      133132        4096   0 Empty
/dev/block/mmcblk0p10         133135      157710      196608   0 Empty
/dev/block/mmcblk0p11         157713      165904       65536  83 Linux
/dev/block/mmcblk0p12         165907      231442      524288  83 Linux
/dev/block/mmcblk0p13         231571      362624     1048432  83 Linux
/dev/block/mmcblk0p14         362753      428288      524288  83 Linux
/dev/block/mmcblk0p15         428417      455016      212800  83 Linux
/dev/block/mmcblk0p16         455169      468992      110592  83 Linux
```

It did work, though, demonstrating that the layout of `partition.dat` does work as detailed in the previous message and can be manipulated successfully.

I think it's likely that I could increase the size of these partitions to give me full use of the entire 8GB, though not on the early partitions where I actually want that space...

**Reply by Diaa19887** [reply]
Welcome Can you send me a file or link to download Security Force so that I can update the Korean screen to the Middle East?

**Comment by adelie93** [comment]
I realized that the offsets were built in to your `gen5.py`.  By changing those offsets to match the ones in my new, modified `partition.dat`, I was successful in getting the system to boot. I increased the size of the `/data` partition successfully.

**Reply by adelie93** [reply]
Then I had the thought: If I can use more of the disk, I should use ALL of the disk (8 GB total), and generated a `partition.dat` that would keep all partitions the same size as original, but increase the `/data` partition to over 4 GB.

This was where things went very badly. I'm not sure why, as all previous tests worked fine. But after flashing this image where I try to use all 8 GB... I now have a brick. 😢

**Reply by rgerganov** [reply]
Do you know if `lk.rom` is corrupted or it is still there? What happens on the screen when you boot holding the buttons?

**Reply by adelie93** [reply]
Nothing happens no matter what I have tried. The screen is completely black, as if the power was removed.

My guess is that `lk.rom` somehow got corrupted. I do not know how that could be possible, though, as I was **very** careful never to modify the first partition so that `lk.rom` should always have been written to the same place.

Maybe there are undocumented things happening during the upgrade process that use the "empty" space at the end of the disk or something?

I tried connecting to the UART as described on the Russian site in the hope that I would find a `u-boot` screen that I could use to recover from, but no luck. Either I did it wrong or there is no `u-boot` there to be found. I have used that technique on other hardware successfully, but not this time.

**Reply by graftonskaggs** [reply]
Has anyone tried to set some of the files in the security Force to 0 kb that way you don't have to update everything and you can just update what you want or does the size have to be the same as it is in that partition table.
